src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -0,0 +1,934 @@
+            return nullptr;
+
+        for (const auto & table : tables)
+        {
+            auto storage = std::get<StoragePtr>(table);
+            const auto & sorting_key = storage->getInMemoryMetadataPtr()->getSortingKey();
+            if (sorting_key.column_names.empty())
+                return nullptr;
+        }
+
+        return merge;
+    }
+
+    return nullptr;
+}
+
+QueryPlan::Node * findReadingStep(QueryPlan::Node & node)
+{
+    IQueryPlanStep * step = node.step.get();
+    if (auto * reading = checkSupportedReadingStep(step))
+        return &node;
+
+    if (node.children.size() != 1)
+        return nullptr;
+
+    if (typeid_cast<ExpressionStep *>(step) || typeid_cast<FilterStep *>(step) || typeid_cast<ArrayJoinStep *>(step))
+        return findReadingStep(*node.children.front());
+
+    if (auto * distinct = typeid_cast<DistinctStep *>(step); distinct && distinct->isPreliminary())
+        return findReadingStep(*node.children.front());
+
+    return nullptr;
+}
+
+/// FixedColumns are columns which values become constants after filtering.
+/// In a query "SELECT x, y, z FROM table WHERE x = 1 AND y = 'a' ORDER BY x, y, z"
+/// Fixed columns are 'x' and 'y'.
+using FixedColumns = std::unordered_set<const ActionsDAG::Node *>;
+
+/// Right now we find only simple cases like 'and(..., and(..., and(column = value, ...), ...'
+void appendFixedColumnsFromFilterExpression(const ActionsDAG::Node & filter_expression, FixedColumns & fixed_columns)
+{
+    std::stack<const ActionsDAG::Node *> stack;
+    stack.push(&filter_expression);
+
+    while (!stack.empty())
+    {
+        const auto * node = stack.top();
+        stack.pop();
+        if (node->type == ActionsDAG::ActionType::FUNCTION)
+        {
+            const auto & name = node->function_base->getName();
+            if (name == "and")
+            {
+                for (const auto * arg : node->children)
+                    stack.push(arg);
+            }
+            else if (name == "equals")
+            {
+                const ActionsDAG::Node * maybe_fixed_column = nullptr;
+                bool is_single = true;
+                for (const auto & child : node->children)
+                {
+                    if (!child->column)
+                    {
+                        if (!maybe_fixed_column)
+                            maybe_fixed_column = child;
+                        else
+                            is_single = false;
+                    }
+                }
+
+                if (maybe_fixed_column && is_single)
+                {
+                    //std::cerr << "====== Added fixed column " << maybe_fixed_column->result_name << ' ' << static_cast<const void *>(maybe_fixed_column) << std::endl;
+                    fixed_columns.insert(maybe_fixed_column);
+
+                    const ActionsDAG::Node * maybe_injective = maybe_fixed_column;
+                    while (maybe_injective->type == ActionsDAG::ActionType::FUNCTION
```

Comment:
Need to add some comment, that we go thought injective functions and it is okay.